javascript/ql/src/Declarations/DeadStoreOfProperty.ql

```diff
@@ -67,67 +95,150 @@ predicate maybeAccessesProperty(Expr e, string name) {
  * Holds if `assign1` and `assign2` both assign property `name`, but `assign1` is dead because of `assign2`.
  */
 predicate isDeadAssignment(string name, DataFlow::PropWrite assign1, DataFlow::PropWrite assign2) {
-  postDominatedPropWrite(name, assign1, assign2) and
-  noPropAccessBetween(name, assign1, assign2) and
+  (
+    assign2.getWriteNode() = getANodeWithNoPropAccessBetweenInsideBlock(name, assign1) and
+    postDominatedPropWrite(name, assign1, assign2, true)
+    or
+    noPropAccessBetweenGlobal(name, assign1, assign2)
+  ) and
   not isDOMProperty(name)
 }
 
 /**
- * Holds if `assign` assigns a property `name` that may be accessed somewhere else in the same block,
+ * Holds if `assign` assigns a property that may be accessed somewhere else in the same block,
  * `after` indicates if the access happens before or after the node for `assign`.
  */
-bindingset[name]
-predicate maybeAccessesAssignedPropInBlock(string name, DataFlow::PropWrite assign, boolean after) {
-  exists(ReachableBasicBlock block, int i, int j, Expr e |
-    assign.getWriteNode() = block.getNode(i) and
-    e = block.getNode(j) and
-    maybeAccessesProperty(e, name)
+predicate maybeAssignsAccessedPropInBlock(DataFlow::PropWrite assign, boolean after) {
+  (
+    // early pruning - manual magic
+    after = true and postDominatedPropWrite(_, assign, _, false)
+    or
+    after = false and postDominatedPropWrite(_, _, assign, false)
+  ) and
+  exists(ReachableBasicBlock block, int i, int j, Expr e, string name |
+    i = getRank(block, assign.getWriteNode(), name) and
+    j = getRank(block, e, name) and
+    isAPropertyRead(e, name)
   |
     after = true and i < j
     or
     after = false and j < i
   )
+  or
+  exists(ReachableBasicBlock block | assign.getWriteNode().getBasicBlock() = block |
+    after = true and isBeforeImpure(assign, block)
+    or
+    after = false and isAfterImpure(assign, block)
+  )
 }
 
 /**
- * Holds if `assign1` and `assign2` both assign property `name`, and the assigned property is not accessed between the two assignments.
+ * Holds if a ControlFlowNode `c` is before an impure expression inside `bb`.
  */
-bindingset[name]
-predicate noPropAccessBetween(string name, DataFlow::PropWrite assign1, DataFlow::PropWrite assign2) {
+predicate isBeforeImpure(DataFlow::PropWrite write, ReachableBasicBlock bb) {
+  getANodeAfterWrite(write, bb).(Expr).isImpure()
+}
+
+/**
+ * Gets a ControlFlowNode that comes after `write` inside `bb`.
+ * Stops after finding the first impure expression
```

Comment:
Here the docstring mentions purity, but it would be nice if the motivation for the purity check could be motivated.
Perhaps the `isBeforeImpure`, `isAfterImpure`, `getANodeAfterWrite`, `getANodeBeforeWrite` should all be wrapped in an internal module, with the latter two as `private` predicates.